src/main/java/io/cdap/plugin/gcp/bigquery/sink/AbstractBigQuerySink.java

```diff
@@ -291,4 +291,36 @@ protected Configuration getOutputConfiguration() throws IOException {
     return configuration;
   }
 
+  /**
+   * Identify a bucket name from the pipeline context and sink configuration
+   * @param context Sink Context
+   * @param config Sink Configuration
+   * @return Bucket name to use for this sink.
+   */
+  protected String getBucketName(BatchSinkContext context, AbstractBigQuerySinkConfig config) {
+    // Get temporary bucket name from sink configuration
+    String bucketName = config.getBucket();
+
+    // Get Bucket Prefix from configuration
+    String bucketPrefix = BigQueryUtil.getBucketPrefix(context.getArguments());
+
+    // If temp bucket name is not defined in configuration, and a common bucket name prefix has been specified,
+    // we must set this prefix along with the destination location in order to create/reuse the bucket.
+    // Otherwise, if temp bucket name is defined, or a prefix is not set, the configureBucket method will prepare
+    // for a new bucket creation.
+    if (bucketName == null && bucketPrefix != null) {
+      // Check if the destination dataset exists.
+      DatasetId datasetId = DatasetId.of(config.getDatasetProject(), config.getDataset());
+      Dataset dataset = bigQuery.getDataset(datasetId);
```

Comment:
It seems like we should already be making this call as part of the existing validation/automatic bucket creation. If so, we should do some refactoring to avoid duplicate calls. It's become kind of a mess now, but ideally we do all the I/O in a single place and pass the return objects around.

Reply:
I don't see an easy way to refactor this without making 2 calls to get the dataset.

The problem is the bucket name is needed even when in preview. This makes it difficult to refactor the code in a way that doesn't change the entire method signature for the abstract bigquery sink and other callers (such as the BigQuery Pushdown implementation).

Reply:
I have refactored the code a bit, hope this helps